pkg/operator/starter.go

```diff
@@ -305,6 +305,7 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		controllerContext.EventRecorder.WithComponentSuffix("invalid-webhook-certs"),
 		"/var/run/configmaps/service-ca-bundle/service-ca.crt",
 		metricscontroller.NewLegacyCNCertsMetricsSyncFunc(
+			"",
 			`sum(apiserver_kube_aggregator_x509_missing_san_total{job="apiserver",namespace="default"}) + sum(apiserver_webhooks_x509_missing_san_total{job="apiserver",namespace="default"})`,
```

Comment:
To give the controller clues about what certificates are faulty let's modify the query to be:
```
sum by (type) (
  label_replace(apiserver_webhooks_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "webhooks", "", "")
  or
  label_replace(apiserver_kube_aggregator_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "aggregation", "", ""))
```

This will yield results like:
```
{type="webhooks"}: 123
{type="aggregation"}: 456
```

Reply:
as discussed OOB let's include a promtool unit test file for this query as well:
```
# This file declares a unit test for the cert detection query. Execute with:
# $ /bin/prometheus-2.x.y/promtool test rules ~/tmp/api.yaml
# Unit Testing:  /Users/surbania/tmp/api.yaml
#  SUCCESS
#
# TODO(sur): add support for promtool in api-build-machinery.

evaluation_interval: 1m
tests:
  - interval: 1m
    input_series:
      #
      # api-server #1 metrics
      #
      - # apiserver #1 started observing invalid SANs at minute 4
        series: 'apiserver_kube_aggregator_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.1:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 0 1'
      - series: 'apiserver_webhooks_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.1:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 0 0'

      #
      # api-server #2 metrics
      #
      - series: 'apiserver_kube_aggregator_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.2:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 0 0'
      - # apiserver #2 started observing invalid SANs at minute 3
        series: 'apiserver_webhooks_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.2:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 1 1'

      #
      # api-server #3 metrics
      #
      - series: 'apiserver_kube_aggregator_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.3:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 0 0'
      - # apiserver #3 started observing invalid SANs at minute 4
        series: 'apiserver_webhooks_x509_missing_san_total{apiserver="kube-apiserver", endpoint="https", instance="10.0.0.3:6443", job="apiserver", namespace="default", service="kubernetes"}'
        values: '0 0 0 0 2'

      #
      # rogue job emitting this metrics, it must be ignored
      #
      - series: 'apiserver_webhooks_x509_missing_san_total{endpoint="https-metrics", instance="10.0.128.4:10250", job="unknonw", metrics_path="/metrics", namespace="foobar", node="ci-op-5fhw80bs-465b4-tfs28-worker-b-db2vd", service="foo"}'
        values: '0 0 0 1 1'

    promql_expr_test:
      - eval_time: '0m'
        expr: 'sum by (type) (
  label_replace(apiserver_webhooks_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "webhooks", "", "")
  or
  label_replace(apiserver_kube_aggregator_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "aggregation", "", ""))'
        exp_samples:
          - labels: '{type="webhooks"}'
            value: 0
          - labels: '{type="aggregation"}'
            value: 0

      - eval_time: '3m'
        expr: 'sum by (type) (
  label_replace(apiserver_webhooks_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "webhooks", "", "")
  or
  label_replace(apiserver_kube_aggregator_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "aggregation", "", ""))'
        exp_samples:
          - labels: '{type="webhooks"}'
            value: 1
          - labels: '{type="aggregation"}'
            value: 0

      - eval_time: '4m'
        expr: 'sum by (type) (
  label_replace(apiserver_webhooks_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "webhooks", "", "")
  or
  label_replace(apiserver_kube_aggregator_x509_missing_san_total{job="apiserver", namespace="default"}, "type", "aggregation", "", ""))'
        exp_samples:
          - labels: '{type="webhooks"}'
            value: 3
          - labels: '{type="aggregation"}'
            value: 1
```